nixos/modules/services/networking/prosody.nix

```diff
@@ -1,887 +1,675 @@
-          message = ''
-            You need to setup the uploadHttp module through
-            config.services.prosody.uploadHttp to comply with
-            XEP-0423.
-          '' + genericErrMsg;}
-      ];
-    in errors;
-
-    environment.systemPackages = [ cfg.package ];
-
-    environment.etc."prosody/prosody.cfg.lua".text =
+    assertions =
       let
-        httpDiscoItems = optionals (cfg.uploadHttp != null)
-            [{ url = cfg.uploadHttp.domain; description = "HTTP upload endpoint";}];
-        mucDiscoItems = builtins.foldl'
-            (acc: muc: [{ url = muc.domain; description = "${muc.domain} MUC endpoint";}] ++ acc)
-            []
-            cfg.muc;
-        discoItems = cfg.disco_items ++ httpDiscoItems ++ mucDiscoItems;
-      in ''
-
-      pidfile = "/run/prosody/prosody.pid"
-
-      log = "*syslog"
-
-      data_path = "${cfg.dataDir}"
-      plugin_paths = {
-        ${lib.concatStringsSep ", " (map (n: "\"${n}\"") cfg.extraPluginPaths) }
-      }
-
-      ${ optionalString  (cfg.ssl != null) (createSSLOptsStr cfg.ssl) }
-
-      admins = ${toLua cfg.admins}
-
-      -- we already build with libevent, so we can just enable it for a more performant server
-      use_libevent = true
-
-      modules_enabled = {
-
-        ${ lib.concatStringsSep "\n  " (lib.mapAttrsToList
-          (name: val: optionalString val "${toLua name};")
-        cfg.modules) }
-        ${ lib.concatStringsSep "\n" (map (x: "${toLua x};") cfg.package.communityModules)}
-        ${ lib.concatStringsSep "\n" (map (x: "${toLua x};") cfg.extraModules)}
-      };
-
-      disco_items = {
-      ${ lib.concatStringsSep "\n" (builtins.map (x: ''{ "${x.url}", "${x.description}"};'') discoItems)}
-      };
-
-      allow_registration = ${toLua cfg.allowRegistration}
-
-      c2s_require_encryption = ${toLua cfg.c2sRequireEncryption}
-
-      s2s_require_encryption = ${toLua cfg.s2sRequireEncryption}
+        checkForModule = mod: b: any (e: e.module == mod) (attrValues b.components);
+        virtualHosts = filterAttrs (_: v: v.settings.enabled) cfg.virtualHosts;
 
-      s2s_secure_auth = ${toLua cfg.s2sSecureAuth}
+        # ensure a given module (e.g. muc or http_upload) is applied or available to every virtual host
+        mkAssertion = module: message: {
+          assertion = cfg.xmppComplianceSuite -> checkForModule module cfg || all (checkForModule module) (attrValues virtualHosts);
+          message = message + ''
 
-      s2s_insecure_domains = ${toLua cfg.s2sInsecureDomains}
+            Having a server not XEP-0423-compliant might make your XMPP
+            experience terrible. See the NixOS manual for further
+            information.
 
-      s2s_secure_domains = ${toLua cfg.s2sSecureDomains}
-
-      authentication = ${toLua cfg.authentication}
-
-      http_interfaces = ${toLua cfg.httpInterfaces}
-
-      https_interfaces = ${toLua cfg.httpsInterfaces}
```

Comment:
These two seem gone from my generated config file; used to be
```
-http_interfaces = { "*", "::" }
-
-https_interfaces = { "*", "::" }
```
(I didn't set interfaces explictly, so the problem would be that while we remap them we don't preserve the default; not sure if important or if prosody default matches)

Reply:
I explicitly tried to dump our defaults when they match upstream in the spirit of RFC42. These are the upstream defaults so all good here 👍